server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -89,112 +293,71 @@ private FetchObjectResponse serialize(String expectedType, SessionState state, O
-            }
-            if (!forceNew) {
-                for (ReferenceImpl reference : references) {
-                    if (equals.test(object, reference.export.get())) {
-                        return Optional.of(reference);
-                    }
+                for (Object reference : references) {
+                    final String type = typeLookup.type(reference).orElse(null);
+                    final ExportObject<?> exportObject = sessionState.newServerSideExport(reference);
+                    exports.add(exportObject);
+                    TypedTicket typedTicket = ticketForExport(exportObject, type);
+                    payload.addExportedReferences(typedTicket);
                 }
-            }
-            return newReferenceImpl(object, allowUnknownType);
-        }
+                final StreamResponse.Builder responseBuilder =
+                        StreamResponse.newBuilder().setData(payload);
 
-        private Optional<Reference> newReferenceImpl(Object object, boolean allowUnknownType) {
-            final String type = typeLookup.type(object).orElse(null);
-            if (!allowUnknownType && type == null) {
-                return Optional.empty();
+                // Explicitly running this unsafely, we want the exception to clean up
+                responseObserver.onNext(responseBuilder.build());
+            } catch (Throwable t) {
+                // Release any exports we failed to send, and report this as a checked exception
+                cleanup(exports, t);
+                throw new ObjectType.ObjectCommunicationException(t);
             }
-            final ExportObject<?> exportObject = sessionState.newServerSideExport(object);
-            final ReferenceImpl ref = new ReferenceImpl(references.size(), type, exportObject);
-            references.add(ref);
-            return Optional.of(ref);
         }
-    }
-
-    private static final class ReferenceImpl implements Reference {
-        private final int index;
-        private final String type;
-        private final ExportObject<?> export;
 
-        public ReferenceImpl(int index, String type, ExportObject<?> export) {
-            this.index = index;
-            this.type = type;
-            this.export = Objects.requireNonNull(export);
-        }
-
-        public TypedTicket typedTicket() {
-            final TypedTicket.Builder builder = TypedTicket.newBuilder().setTicket(export.getExportId());
+        private TypedTicket ticketForExport(ExportObject<?> exportObject, String type) {
+            TypedTicket.Builder builder = TypedTicket.newBuilder().setTicket(exportObject.getExportId());
             if (type != null) {
                 builder.setType(type);
             }
             return builder.build();
         }
 
         @Override
-        public int index() {
-            return index;
+        public void onClose() {
+            closed = true;
```

Comment:
Should this be `volatile`?

Reply:
Jianfeng and I have been talking about other thread safety concerns, I'm going to change the following line (onCompleted) to use our safelyComplete util method (which is synchronized on the response stream to ensure we don't race with ourselves). This change also needs to be made when responses are sent to the client (this time just wrapping with a synchronized block, as GrpcUtil methods would do, but we actually want exceptions to escape to the caller). I think that resolves this?

Note that this field only exists to detect in-thread closing for using the legacy fetch-only gRPC method, and once we remove it, can be safely deleted.

Reply:
I think this is resolved, PTAL